File: tests/fixtures/lang_samples/go/perf_io_in_loop.go
// Fixture for the Go performance dialect (io_in_loop / string_concat /
// defer_in_loop / regex_compile_in_loop). Every POSITIVE is a genuine
// per-iteration cost; every NEGATIVE is a shape the dialect must NOT flag.
// Hand-counted expectations live in test_perf_go.py. NOTE: this file imports
// database/sql, so file-level db evidence is present; ambiguous-verb gating
// (GORM Find without a db import) is tested inline instead.
package sample

import (
	"database/sql"
	"net/http"
	"os"
	"regexp"
)

// --- POSITIVES -------------------------------------------------------------

func dbQueryInRange(db *sql.DB, ids []int) {
	for _, id := range ids {
		db.Query("SELECT 1", id) // POSITIVE: database/sql round-trip per item
	}
}

func httpGetInRange(urls []string) {
	for _, u := range urls {
		http.Get(u) // POSITIVE: network per item
	}
}

func osOpenInRange(paths []string) {
	for _, p := range paths {
		os.Open(p) // POSITIVE: filesystem per item
	}
}

func deferInLoop(paths []string) {
	for _, p := range paths {
		f, _ := os.Open(p) // POSITIVE: filesystem (also the defer below)
		defer f.Close()    // POSITIVE: defer in loop leaks the handle
	}
}

func regexpCompileInRange(ids []string) {
	for _, id := range ids {
		regexp.MustCompile(`^[a-z]+$`) // POSITIVE: a static literal pattern, hoistable
		regexp.MustCompile(id)         // NEGATIVE: a dynamic per-iteration arg cannot be hoisted
	}
}

func stringConcatInRange(rows []string) string {
	s := ""
	for range rows {
		s += "row" // POSITIVE: quadratic string build (literal RHS)
	}
	return s
}

// --- NEGATIVES -------------------------------------------------------------

func dbQueryConstantLoop(db *sql.DB) {
	for i := 0; i < 10; i++ { // NEGATIVE: constant-bound three-clause loop
		db.Query("SELECT 1")
	}
}

func pureCallInRange(items []int) {
	total := 0
	for _, x := range items {
		total += x // NEGATIVE: numeric +=, no I/O
	}
	_ = total
}

func queryOutsideLoop(db *sql.DB) {
	db.Query("SELECT 1") // NEGATIVE: runs once, not in a loop
}

File: tests/fixtures/lang_samples/java/PerfIoInLoop.java
// Fixture for the Java performance dialect (io_in_loop / string_concat /
// regex_compile_in_loop). Every POSITIVE is a genuine per-iteration I/O
// boundary or wasted recompile; every NEGATIVE is a shape the dialect must NOT
// flag. Hand-counted expectations live in test_perf_java.py. NOTE: this file
// imports a db library, so file-level db evidence is present; ambiguous-verb
// gating (find/get without a db import) is tested inline instead.
import java.nio.file.Files;
import java.util.List;
import java.util.regex.Pattern;
import org.springframework.data.jpa.repository.JpaRepository;

class PerfIoInLoop {

    // --- POSITIVES ---------------------------------------------------------

    void springDataN1(List<String> ids) {
        for (String id : ids) {
            repo.findByName(id);            // POSITIVE: Spring-Data derived query
        }
    }

    void jdbcInLoop(List<String> ids, java.sql.Statement st) {
        for (String id : ids) {
            st.executeQuery(id);            // POSITIVE: JDBC round-trip
        }
    }

    void filesInLoop(List<String> paths) {
        for (String p : paths) {
            Files.readString(java.nio.file.Path.of(p));  // POSITIVE: filesystem
        }
    }

    void newStreamInLoop(List<String> paths) {
        for (String p : paths) {
            new java.io.FileInputStream(p);  // POSITIVE: filesystem constructor
        }
    }

    void regexCompileInLoop(List<String> ids) {
        for (String id : ids) {
            Pattern.compile(id);             // POSITIVE: recompiled per iteration
        }
    }

    String stringConcatInLoop(List<String> rows) {
        String out = "";
        for (String r : rows) {
            out += "line";                   // POSITIVE: quadratic string build
        }
        return out;
    }

    // --- NEGATIVES ---------------------------------------------------------

    void pureCallInLoop(List<String> items) {
        int total = 0;
        for (String x : items) {
            total += x.length();             // NEGATIVE: numeric +=, pure call
        }
    }

    void sinkOutsideLoop(java.sql.Statement st) {
        st.executeQuery("SELECT 1");         // NEGATIVE: runs once, not in a loop
    }

    void regexCompileOnce(List<String> ids) {
        Pattern p = Pattern.compile("[0-9]+"); // NEGATIVE: compiled once, hoisted
        for (String id : ids) {
            p.matcher(id).matches();           // NEGATIVE: matches() is not a sink
        }
    }
}
